### PR TITLE
[#1810] Allow extensions autoloading in osqueryi

### PR DESCRIPTION
**Pull request by theopolis**
The osquery shell can autoload extensions. The only difference between the daemon and the shell, with respect to extensions autoloading, is the daemon expects to fork a worker process and continue only as a watch dog (a loop to check sanity of a single worker). If the shell implicitly disable watchdog features it can offer the same extension loading and monitoring as the daemon.

### CODE REVIEW

**Comment by osqueryer** [comment]
Build origin/pr/1836/merge (Job: 2369) is good! :+1:

**Comment by osqueryer** [comment]
Build origin/pr/1836/merge (Job: 2370) is good! :+1:

File: osquery/extensions/extensions.cpp [hunk 1]
```diff
@@ -89,17 +91,30 @@ void ExtensionWatcher::exitFatal(int return_code) {
...
+  // This does NOT use pingExtension to avoid the latency checks applied.
   ExtensionStatus status;
-  try {
-    auto client = EXManagerClient(path_);
-    // Ping the extension manager until it goes down.
-    client.get()->ping(status);
-  } catch (const std::exception& e) {
-    LOG(WARNING) << "Extension watcher ending: osquery core has gone away";
+  bool core_sane = true;
+  if (isWritable(path_)) {
```
Comment (theopolis) [inline osquery/extensions/extensions.cpp]: This check helps prevent spurious warnings in older version of Thrift. There is still a tear-down race within exceptions when aborting their thread pool in response to a killed core/watcher. Solving this race is not high priority at the moment.

File: osquery/registry/registry.cpp [hunk 2]
```diff
@@ -60,14 +60,21 @@ Status RegistryHelperCore::setActive(const std::string& item_name) {
...
   for (const auto& item : osquery::split(item_name, ",")) {
     if (exists(item, true)) {
-      Registry::get(name_, item)->setUp();
+      status = Registry::get(name_, item)->setUp();
+    } else if (exists(item, false) && !Registry::external()) {
+      // If the active plugin is within an extension we must wait.
+      // An extension will first broadcast the registry, then receive the list
+      // of active plugins, active them if they are extension-local, and finally
+      // start their extension socket.
+      status = pingExtension(getExtensionSocket(external_.at(item_name)));
```
Comment (theopolis) [inline osquery/registry/registry.cpp]: This seems to be the source of a few races. There was an incorrect assumption that if the plugin appeared in the core's registry, then it was acceptable to proceed assuming the extension's registry was setup. The correct flow requires a check for a valid extension socket. Extensions will only start their receive sockets when their registry is setup. The core however, receives an extension's broadcast before assigning it a UUID.

File: osquery/registry/registry.cpp [hunk 3]
```diff
@@ -60,14 +60,21 @@ Status RegistryHelperCore::setActive(const std::string& item_name) {
     }
   }
 
+  Status status(0, "OK");
   active_ = item_name;
   // The active plugin is setup when initialized.
   for (const auto& item : osquery::split(item_name, ",")) {
     if (exists(item, true)) {
-      Registry::get(name_, item)->setUp();
+      status = Registry::get(name_, item)->setUp();
```
Comment (theopolis) [inline osquery/registry/registry.cpp]: I'm a tad worried about this new status dependency but believe it's the best pattern moving forward. If a config or logger fails to setup and it is the active plugin then the core/daemon/shell should respond appropriately and exit.

**Pull request merged**